infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -364,15 +364,58 @@ protected String getNamespaceName(Workspace workspace) throws InfrastructureExce
               EnvironmentContext.getCurrent().getSubject().getUserId(),
               EnvironmentContext.getCurrent().getSubject().getUserName());
       namespace = evaluateLegacyNamespaceName(resolutionCtx);
+
+      LOG.warn(
+          "Workspace '{}' doesn't have an explicit namespace assigned."
+              + " The legacy namespace resolution resolved it to '{}'.",
+          workspace.getId(),
+          namespace);
     }
+
+    if (!NamespaceNameValidator.isValid(namespace)) {
+      // At a certain unfortunate past version of Che, we stored invalid namespace names.
+      // At this point in time, we're trying to work with an existing workspace that never could
+      // started OR has been running since before that unfortunate version. In both cases, going
+      // back to the default namespace name is the most safe bet we can make.
+
+      // but of course, our attempt will be futile if we're running in a context that doesn't know
+      // the current user.
+      Subject subj = EnvironmentContext.getCurrent().getSubject();
+      if (!subj.isAnonymous()) {
+        NamespaceResolutionContext resolutionCtx =
+            new NamespaceResolutionContext(workspace.getId(), subj.getUserId(), subj.getUserName());
+
+        String defaultNamespace = evaluateNamespaceName(resolutionCtx);
+
+        LOG.warn(
+            "The namespace '{}' of the workspace '{}' is not valid. Trying to recover"
+                + " from this situation using a default namespace which resolved to '{}'.",
+            namespace,
+            workspace.getId(),
+            defaultNamespace);
+
+        namespace = defaultNamespace;
+      } else {
+        // log a warning including a stacktrace to be able to figure out from where we got here...
+        LOG.warn(
+            "The namespace '{}' of the workspace '{}' is not valid but we currently don't have"
+                + " an active user to try an recover from this situation.",
+            namespace,
+            workspace.getId(),
+            new Exception());
```

Comment:
this exception will be very confusing. At least I would use `Throwable` instead of `Exception` and write message there something like `this is here just to see the stacktrace`.

You commented you need stacktrace, so better log just stacktrace with `Thread.currentThread().getStacktrace()`.

Reply:
The exception is used as a parameter to the log statement and therefore the log message will be followed by the exception stacktrace in the log output.

How do I log an array of StackTraceElement`s in a single line like this?